zebra-chain/src/parameters/network.rs

```diff
@@ -279,3 +275,84 @@ impl FromStr for Network {
 #[derive(Clone, Debug, Error)]
 #[error("Invalid network: {0}")]
 pub struct InvalidNetworkError(String);
+
+impl zp_consensus::Parameters for Network {
+    fn activation_height(
+        &self,
+        nu: zcash_primitives::consensus::NetworkUpgrade,
+    ) -> Option<zcash_primitives::consensus::BlockHeight> {
+        let target_nu = match nu {
+            zp_consensus::NetworkUpgrade::Overwinter => NetworkUpgrade::Overwinter,
+            zp_consensus::NetworkUpgrade::Sapling => NetworkUpgrade::Sapling,
+            zp_consensus::NetworkUpgrade::Blossom => NetworkUpgrade::Blossom,
+            zp_consensus::NetworkUpgrade::Heartwood => NetworkUpgrade::Heartwood,
+            zp_consensus::NetworkUpgrade::Canopy => NetworkUpgrade::Canopy,
+            zp_consensus::NetworkUpgrade::Nu5 => NetworkUpgrade::Nu5,
+        };
+
+        // Heights are hard-coded below Height::MAX or checked when the config is parsed.
+        target_nu
```

Comment:
This is called at least twice for every valid Sapling output in coinbase transactions (`zcash_note_encryption::try_sapling_output_recovery()` eventually calls `plaintext_version_is_valid`). It's slightly worse for configured activation heights where some network upgrades are omitted, because it'll iterate through the list for every network upgrade after `self`.

I think this is fine because:
- It doesn't seem significant relative to the decryption and key derivations in `try_sapling_output_recovery()`
- There should be few cache-misses if it's called frequently since it's the same memory location across calls
- It should only be ~40B so it won't take up much space in the cache, and
- Semantic block validation is done in parallel and sync performance is likely constrained by the sequential contextual validation in the state service write task

So it seems unlikely to measurably affect sync performance for most users.

We could open an issue to benchmark this and other factors in sync performance, if the performance impact _is_ measurable, we could use a struct instead (like the one used for configuring activation heights).